libs/s25main/lua/LuaInterfaceGame.cpp

```diff
@@ -313,6 +319,42 @@ void LuaInterfaceGame::PostMessageWithLocation(int playerIdx, const std::string&
                                                       gw.MakeMapPoint(Position(x, y))));
 }
 
+namespace {
+void MarkCampaignChapter(const std::string& campaignUid, unsigned char chapter, char code)
+{
+    if(chapter < 1)
+        return;
+
+    auto& saveData = SETTINGS.campaigns.saveData[campaignUid];
+
+    if(saveData.length() < chapter)
+        saveData.resize(chapter);
+
+    auto& chapterSaveData = saveData[chapter - 1];
+
+    if(chapterSaveData == CampaignSaveCodes::chapterCompleted)
+        return;
+
+    chapterSaveData = code;
```

Comment:
This is more clear than an "early" return

```suggestion
    if(chapterSaveData != CampaignSaveCodes::chapterCompleted)
        chapterSaveData = code;
```